src/services/action/repositories/action.test.ts

```diff
@@ -0,0 +1,738 @@
+import {
+  ActionFactory,
+  AggregateBy,
+  AggregateFunction,
+  AggregateMetric,
+  Context,
+  CountGroupBy,
+  DiscriminatedItem,
+} from '@graasp/sdk';
+
+import build, { clearDatabase } from '../../../../test/app';
+import { AppDataSource } from '../../../plugins/datasource';
+import { saveItem } from '../../item/test/fixtures/items';
+import { saveMember } from '../../member/test/fixtures/members';
+import { DEFAULT_ACTIONS_SAMPLE_SIZE } from '../constants/constants';
+import { Action } from '../entities/action';
+import { expectActions } from '../test/fixtures/actions';
+import { ActionRepository } from './action';
+
+// mock datasource
+jest.mock('../../../plugins/datasource');
+
+const rawRepository = AppDataSource.getRepository(Action);
+
+describe('Action Repository', () => {
+  let app;
+  let actor;
+  let member;
+  let item;
+
+  beforeEach(async () => {
+    ({ app, actor } = await build());
+    member = await saveMember();
+
+    item = await saveItem({ actor });
+  });
+
+  afterEach(async () => {
+    jest.clearAllMocks();
+    await clearDatabase(app.db);
+    actor = null;
+    member = null;
+    item = null;
+    app.close();
+  });
+
+  describe('postMany', () => {
+    it('save many actions', async () => {
+      const actions = [
+        ActionFactory({ item: null, member: null }),
```

Comment:
Yes because otherwise it creates random item and member, that are not in the database ^^' It could make sense to have a util function to save actions correctly tho. What do you think?

Reply:
Okay, that's clearer to me; thanks for your reply! 😄

I'm thinking about the `ActionFactory`, specifically regarding the scenario where no item is set, prompting it to generate one. In my view, I expect the `ActionFactory` to produce a valid action. However, what happens if the action contains an invalid (non-existent) item or member? Is the generated action still considered valid?

I don't have a definitive solution yet; perhaps I'm mistaken, and it's entirely acceptable to generate a new action with an item that isn't yet in the database...

Perhaps the `ActionFactory` should only concern itself with the data that the action manages. This means that when obtaining a new action from the factory, it would come with item and member **set to null**. Consequently, if you require an action with a specific item and/or member, you would pass those parameters to the factory. If the items don't exist, the responsibility doesn't lie with the `ActionFactory`, but rather with the user, who **must ensure that the foreign entity is correctly set in the database** before trying to save the created action.

In this manner, if the user desires an action with a newly generated item, they would simply utilize the ItemFactory, which is responsible for generating a valid item (meaning it can be sent to the repo for saving without triggering exceptions). Another option that arises here could be, as you suggested, to introduce utility functions tasked with generating and saving entities in the correct order to ensure that the generated action can be successfully saved to the database (if not already done by the utility function).

I would be interested to have your opinions about that @pyphilia and @spaenleh 😆.